azure-iot-device/azure/iot/device/iothub/pipeline/http_pipeline.py

```diff
@@ -0,0 +1,126 @@
+# --------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import sys
+from azure.iot.device.common.evented_callback import EventedCallback
+from azure.iot.device.common.pipeline import (
+    pipeline_stages_base,
+    pipeline_ops_base,
+    pipeline_stages_http,
+)
+
+from azure.iot.device.iothub.pipeline import exceptions as pipeline_exceptions
+
+from . import (
+    constant,
+    pipeline_stages_iothub,
+    pipeline_ops_iothub,
+    pipeline_ops_iothub_http,
+    pipeline_stages_iothub_http,
+)
+from azure.iot.device.iothub.auth.x509_authentication_provider import X509AuthenticationProvider
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPPipeline(object):
+    """Pipeline to communicate with Edge.
+    Uses HTTP.
+    """
+
+    def __init__(self, auth_provider, pipeline_configuration):
+        """
+        Constructor for instantiating a pipeline adapter object
+        :param auth_provider: The authentication provider
+        :param pipeline_configuration: The configuration generated based on user inputs
+        """
+        self._pipeline = (
+            pipeline_stages_base.PipelineRootStage(pipeline_configuration=pipeline_configuration)
+            .append_stage(pipeline_stages_iothub.UseAuthProviderStage())
+            .append_stage(pipeline_stages_iothub_http.IoTHubHTTPTranslationStage())
+            .append_stage(pipeline_stages_http.HTTPTransportStage())
+        )
+
+        callback = EventedCallback()
+
+        if isinstance(auth_provider, X509AuthenticationProvider):
+            op = pipeline_ops_iothub.SetX509AuthProviderOperation(
+                auth_provider=auth_provider, callback=callback
+            )
+        else:  # Currently everything else goes via this block.
+            op = pipeline_ops_iothub.SetAuthProviderOperation(
+                auth_provider=auth_provider, callback=callback
+            )
+
+        self._pipeline.run_op(op)
+        callback.wait_for_completion()
+
+    def invoke_method(self, device_id, method_params, callback, module_id=None):
+        """
+        Send a method response to the service.
+        """
```

Comment:
Here and everywhere in this file: These methods should have docstrings that spell out the errors that can be returned (even though it's via the callback parameter) - it's important for our own documentation to know what can possibly escape this layer.